File: preprocess_2D_data/segmentationAlgorithm2.m
function new_mask = segmentationAlgorithm2(x)
CC = bwconncomp(x > 0);
numPixels = cellfun(@numel,CC.PixelIdxList);
[biggest,idx] = max(numPixels);
mask = x*0;
mask(CC.PixelIdxList{idx}) = 1;
fmask = flip(mask, 2);

symmask = (mask + flip(mask, 2)) > 0;
se = strel('disk',3);

dmask = imerode(symmask, se);
CC = bwconncomp(dmask);
numPixels = cellfun(@numel,CC.PixelIdxList);
[biggest,idx] = max(numPixels);
new_mask = dmask * 0;
new_mask(CC.PixelIdxList{idx}) = 1;

se5 = strel('disk',5);
% reverse dilation
new_mask = imdilate(new_mask, se5);
new_mask = imfill(new_mask, 'holes');
end 

File: preprocess_2D_data/segmentationAlgorithm2.py
import numpy as np
from skimage.morphology import disk, erosion, dilation
from skimage.measure import label, regionprops
from skimage.segmentation import clear_border
import matplotlib.pyplot as plt
from skimage import io

def segmentation_algorithm2(x):
    # Find connected components
    labeled_array, num_features = label(x > 0, return_num=True)
    region_props = regionprops(labeled_array)
    
    # Find the largest connected component
    largest_region = max(region_props, key=lambda region: region.area)
    mask = np.zeros_like(x)
    mask[labeled_array == largest_region.label] = 1
    
    # Flip the mask horizontally
    fmask = np.flip(mask, axis=1)
    
    # Create a symmetric mask
    symmask = (mask + fmask) > 0
    
    # Create a structural element
    se = disk(3)
    
    # Erode the symmetric mask
    dmask = erosion(symmask, se)
    
    # Find connected components in the eroded mask
    labeled_array, num_features = label(dmask, return_num=True)
    region_props = regionprops(labeled_array)
    
    # Find the largest connected component in the eroded mask
    largest_region = max(region_props, key=lambda region: region.area)
    new_mask = np.zeros_like(dmask)
    new_mask[labeled_array == largest_region.label] = 1
    
    # Create a larger structural element for dilation
    se5 = disk(5)
    
    # Dilate the mask
    new_mask = dilation(new_mask, se5)
    
    # Fill holes in the mask
    new_mask = clear_border(new_mask)
    
    return new_mask
